prebuild/emscripten/post.js

```diff
@@ -0,0 +1,1010 @@
+};;
+
+P2.prototype['encode_to'] = P2.prototype.encode_to = /** @suppress {undefinedVars, duplicate} @this{Object} */function(msg, DST, aug) {
+  var self = this.ptr;
+  ensureCache.prepare();
+
+  const [_dst] = ensureRef(DST);
+  const [_msg, msg_len] = ensureRef(msg);
+  const [_aug, aug_len] = ensureRef(aug);
+
+  return wrapPointer(_emscripten_bind_P2_encode_to_5(self, _msg, msg_len, _dst, _aug, aug_len), P2);
+};;
+
+P2.prototype['mult'] = P2.prototype.mult = /** @suppress {undefinedVars, duplicate} @this{Object} */function(scalar) {
+  var self = this.ptr;
+  const [_scalar, scalar_len] = ensureRef(scalar);
+  return wrapPointer(_emscripten_bind_P2_mult_1(self, _scalar, scalar_len * 8), P2);
+};;
+
+P2.prototype['neg'] = P2.prototype.neg = /** @suppress {undefinedVars, duplicate} @this{Object} */function() {
+  var self = this.ptr;
+  return wrapPointer(_emscripten_bind_P2_neg_0(self), P2);
+};;
+
+P2.prototype['cneg'] = P2.prototype.cneg = /** @suppress {undefinedVars, duplicate} @this{Object} */function(flag) {
+  var self = this.ptr;
+  if (flag && typeof flag === 'object') flag = flag.ptr;
+  return wrapPointer(_emscripten_bind_P2_cneg_1(self, flag), P2);
+};;
+
+P2.prototype['add'] = P2.prototype.add = /** @suppress {undefinedVars, duplicate} @this{Object} */function(a) {
+  var self = this.ptr;
+  switch(a.constructor) {
+    case P2:
+      return wrapPointer(_emscripten_bind_P2_add_1__P2(self, a.ptr), P2);
+    case P2_Affine:
+      return wrapPointer(_emscripten_bind_P2_add_1__P2_Affine(self, a.ptr), P2);
+    default:
+      throw new Error("unsuported type for `a` arg");
+  }
+};;
+
+P2.prototype['dbl'] = P2.prototype.dbl = /** @suppress {undefinedVars, duplicate} @this{Object} */function() {
+  var self = this.ptr;
+  return wrapPointer(_emscripten_bind_P2_dbl_0(self), P2);
+};;
+
+  P2.prototype['__destroy__'] = P2.prototype.__destroy__ = /** @suppress {undefinedVars, duplicate} @this{Object} */function() {
+  var self = this.ptr;
+  _emscripten_bind_P2___destroy___0(self);
+};
+// SecretKey
+/** @suppress {undefinedVars, duplicate} @this{Object} */function SecretKey() {
+  this.ptr = _emscripten_bind_SecretKey_SecretKey_0();
+  getCache(SecretKey)[this.ptr] = this;
+};;
+SecretKey.prototype = Object.create(WrapperObject.prototype);
+SecretKey.prototype.constructor = SecretKey;
+SecretKey.prototype.__class__ = SecretKey;
+SecretKey.__cache__ = {};
+Module['SecretKey'] = SecretKey;
+
+SecretKey.prototype['keygen'] = SecretKey.prototype.keygen = /** @suppress {undefinedVars, duplicate} @this{Object} */function(ikm, info) {
+  var self = this.ptr;
+  var ikm_len = 0;
+  ensureCache.prepare();
+  if (ikm && typeof ikm === 'object') {
+    if (ikm instanceof Uint8Array) {
+      ikm_len = ikm.length;
+      ikm = ensureUint8(ikm, ikm.length);
+    } else {
+      throw new Error("unsupported ikm type");
+    }
+  } else {
+    ikm_len = ikm.length;
+    ikm = ensureString(ikm);
```

Comment:
Formally speaking the `ikm_len` can be deceptive in the context, String. The length method returns amount of characters, while what should reach C is array of bytes. The lengths of the two are not the same if there are non-ASCII characters. BTW, I don't see the `intArrayFromString` implementation and it doesn't seem to be standard. Where is it? In some dependency? Just in case, expectation is that String would be encoded as UTF-8...

Reply:
> I don't see the `intArrayFromString` implementation

It's provided by emscripten as part for the current module interface, right? As for `ikm_len`, I gather it should be `ikm_len = lengthBytesUTF8(ikm);`

Reply:
`intArrayFromString` is part of [emscripten's preamble.js](https://emscripten.org/docs/api_reference/preamble.js.html#intArrayFromString) API. When emcc builds the WASM module, it also dumps the preamble into the accompanying JavaScript (before dead-code elimination).

>Just in case, expectation is that String would be encoded as UTF-8...
Correct, is there a potential pitfall there?

Reply:
> > Just in case, expectation is that String would be encoded as UTF-8...

This came out wrong. Expectation is that all strings, not "String" as in "Javascript String", that are given to blst would be encoded as UTF-8. And the trouble is that "Javascript String"-s are stored as UTF-16. intArrayFromString does the UTF-16->UTF-8 conversion, but lengths' meanings are different. On the Javascript side it's amount of UTF-16 characters, while on blst side is the amount of bytes it takes to encode the string in question in UTF-8. To give an example. In Javascript `"Straße".length` is 6, while the amount of bytes corresponding UTF-8 string takes is 7.

Reply:
Understood, thanks for clarifying :+1: